### PR TITLE
Disallowed reuse of data records

### PR DESCRIPTION
**Pull request by liyinan926**
Currently Etractor.readRecord() takes an instance of the data record class, which gives the Etractor chances to reuse the same data record. Originally, the purpose of having this was for performance reason. However, extracted data records now go to the forks through a blocking record queue, which may hold multiple references to the same data records if the writer is slower than the extractor. Allowing data record reuse will lead to undesired or even unexpected behavior if a data record is mutated while the queue is still holding more than one references to it. This commit changed the Extractor interface to disallow data record reuse.

This commit will break `Extractor` implementation classes in other repos. It will also break classes that extend the following classes that have been renamed in this commit:
1. `HadoopExtractor` => `AvroFileExtractor`.
2. `HadoopSource` => `AvroFileSource`.
3. `DatePartitionedExtractor` => `DatePartitionedAvroFileExtractor`.

Signed-off-by: Yinan Li <liyinan926@gmail.com>

### CODE REVIEW

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFileSource.java [hunk 1]
```diff
@@ -28,13 +28,13 @@
 import gobblin.source.extractor.filebased.FileBasedSource;
 
 
-public class HadoopSource extends FileBasedSource<Schema, GenericRecord> {
-  private Logger log = LoggerFactory.getLogger(HadoopSource.class);
+public class AvroFileSource extends FileBasedSource<Schema, GenericRecord> {
+  private Logger log = LoggerFactory.getLogger(AvroFileSource.class);
```
Comment (zliu41) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFileSource.java]: static

Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFileSource.java]: Fixed.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFileExtractor.java [hunk 2]
```diff
@@ -11,33 +11,34 @@
...
   @Override
-  public Iterator<D> downloadFile(String file)
+  public Iterator<GenericRecord> downloadFile(String file)
       throws IOException {
-    DataFileReader<GenericRecord> dfr = null;
     try {
-      dfr = ((HadoopFsHelper) this.fsHelper).getAvroFile(file);
-      fileHandles.put(file, dfr);
-      return (Iterator<D>) dfr;
+      return this.closer.register(((HadoopFsHelper) this.fsHelper).getAvroFile(file));
```
Comment (zliu41) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFileExtractor.java]: rename `HadoopFsHelper` to `AvroFileHelper`?

Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFileExtractor.java]: Yes, good point.

**Comment by liyinan926** [comment]
Updated https://github.com/liyinan926/gobblin/commit/89a65949ffee46c49e552b67eff1201236e0c50a.

**Comment by liyinan926** [comment]
@sahilTakiar, can you also take a look at this pull request? This will break the purger code due to the API change and renaming of some classes.

File: gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsFsHelper.java [hunk 3]
```diff
@@ -40,17 +40,19 @@
 import gobblin.configuration.State;
 
 
-public class HadoopFsHelper implements FileBasedHelper {
-  private static Logger log = LoggerFactory.getLogger(HadoopFsHelper.class);
+public class AvroFsFsHelper implements FileBasedHelper {
```
Comment (sahilTakiar) [inline gobblin-core/src/main/java/gobblin/source/extractor/hadoop/AvroFsFsHelper.java]: Should just be AvroFsHelper instead of AvroFsFsHelper

**Comment by zliu41** [comment]
LGTM barring the typo

**Comment by liyinan926** [comment]
Typo fixed and updated.

**Comment by liyinan926** [comment]
Updated after a discussion with @chavdar. We will still keep the current API, but marking the input parameter `reuse` of `readRecord` as `@Deprecated` and making it clear in the Javadoc. When `Task` calls this method, it will pass in a `null`, which essentially force a new object to be created.

File: gobblin-core/src/main/java/gobblin/source/extractor/filebased/FileBasedExtractor.java [hunk 4]
```diff
@@ -41,32 +40,32 @@
...
+public abstract class FileBasedExtractor<S, D> implements Extractor<S, D> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(FileBasedExtractor.class);
+
+  protected final WorkUnit workUnit;
+  protected final WorkUnitState workUnitState;
+  protected final FileBasedHelper fsHelper;
+  protected final List<String> filesToPull;
+
+  protected final Closer closer = Closer.create();
```
Comment (sahilTakiar) [inline gobblin-core/src/main/java/gobblin/source/extractor/filebased/FileBasedExtractor.java]: Can closer objects be re-used? e.g. can you call closer.register(), closer.close(), and then closer.register() again? If not then we will have to create a new closer instance for each file downloaded.

Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/source/extractor/filebased/FileBasedExtractor.java]: Yes, internally `Closer` just maintains a stack of `Closeable`s. `register` simply pushes the new `Closeable` onto the stack, and `close` simply closes all `Closeable`s on the stack.

Comment (sahilTakiar) [inline gobblin-core/src/main/java/gobblin/source/extractor/filebased/FileBasedExtractor.java]: Ok sounds good.

**Comment by sahilTakiar** [comment]
LGTM

**Comment by liyinan926** [comment]
@sahilTakiar, please update the tracking purger once I merge this on.

**Comment by sahilTakiar** [comment]
Will do

**Pull request merged**